plugins/catalog/src/hooks/useEntities.ts

```diff
@@ -40,6 +40,88 @@ type UseEntities = {
   selectTypeFilter: (id: string) => void;
 };
 
+type EntityFilterGroupOutput<T> = {
+  selectItems: (items: string[]) => void;
+  filteredItems: T[];
+  states: OutputState;
+};
+
+type OutputState = { [key: string]: { isSelected: boolean; count: number } };
+
+type FilterDefinition<T> = {
+  [key: string]: {
+    isSelected: boolean;
+    filterFunction: (entity: T) => boolean;
+  };
+};
+
+export const useEntityFilterGroup = <T>(
+  entities: T[],
+  filterFunctions: FilterDefinition<T>,
+): EntityFilterGroupOutput<T> => {
+  const [filterFuncs, setFilterFuncs] = useState<FilterDefinition<T>>(
```

Comment:
so at this point i think we'll have `const [selectedItems, setSelectedItems] = useState<string[]>([])`, and setSelectedItems is basically what's returned below in selectItems.

then `const allEntities = useContext(EntityFilterGroupsContext).entities`

then there will be a `const filteredEntities = useMemo(() => { allEntites.filter(...) }, [allEntities, selectedItems])`